### ISSUE TITLE
[BUG] checkm vs anvio completeness discrepancy that goes beyond what one would expect

### ISSUE THREAD

**Issue by vdenef** [merenlab/anvio#1954]
## Short description of the problem

Used the same workflow, in one case the checkm and anvio data corresponded, in other case they did not

## anvi'o version

Anvi'o .......................................: hope (v7)

Profile database .............................: 35
Contigs database .............................: 20
Pan database .................................: 14
Genome data storage ..........................: 7
Auxiliary data storage .......................: 2
Structure database ...........................: 2
Metabolic modules database ...................: 2
tRNA-seq database ............................: 1

## System info

using the umich hpc system. The Great Lakes cluster runs CentOS 7. We update the operating system on Great Lakes as CentOS releases new versions and our library of third-party applications offers support. Due to the need to support several types of drivers (AFS and Lustre file system drivers, InfiniBand network drivers and NVIDIA GPU drivers) and dozens of third party applications, we are cautious in upgrading and can lag CentOS’s releases by months. Anvio installed by the system admins, so not sure about all the details here.

## Detailed description of the issue

Started from one single sample assembly.

Workflow 1 mapped 2 samples to this assembly and binned with concoct, then created the anvio profile and contig databases for manual bin refinement. checkm and anvio data corresponded well. 

Workflow 2 mapped these same 2 samples, and an additional 6 other ones from the same environment to the same assembly, binned with concoct, then created the anvio profile and contig databases for manual bin refinement. While the bin size and number of contigs / splits correspond between the concoct bins analyzed by checkm and the ones in the anvio database, the checkm and anvio completeness and redundancy data are off for all bins. For example, checkm says 100% complete and 200% redundancy, and anvio says 50% complete and 25% redundancy. This is across the board.  

I've had issues like this before, but the it traced back to something that got messed up in the contigs renaming and then the bins in checkm had different total lengths than in anvi'o but this does not seem to be the case here.

## Files to reproduce
I add the merged profile.db, the contigs.db, and the checkm info in this folder: (note that the bin numbers in the checkm file, which are just numbers, are named bin_xxx in the anvio database)
https://drive.google.com/drive/folders/1ipv3A-CJ4vg1yfzGdpWemC3l0ZvgL0Ag?usp=sharing

Thanks!
Vincent

**Comment by meren** [comment]
TL;DR: you simply need to regenerate your contigs-db for this project, unfortunately :( I was taking notes while I was trying to diagnose this problem, and I will share them as is so you have a deeper understanding if you would like. Please remove access to the data-pack through Google Drive, and thank you again for sharing a reproducible example.

---

To investigate this weird issue, I randomly picked a bin while going through the CheckM results:

|**Bin Id**|**Marker lineage**|**# genomes**|**# markers**|**# marker sets**|**0**|**1**|**2**|**3**|**4**|**5+**|**Completeness**|**Contamination**|**Strain heterogeneity**|
|:--|:--|:--|:--|:--|:--|:--|:--|:--|:--|:--|:--|:--|:--|
|198|k__Bacteria (UID203)|5449|104|58|0|14|89|1|0|0|100.00|86.76|4.35|

A bin with a high completion and high redundancy estimate. GOOD START.

I run `anvi-split` to get that bin into its own little project:

```
anvi-split -p merged_profile_dbs/ND-57/PROFILE.db \
                 -c contig_dbs/ND-57-CONTIGS.db \
                 -C concoct \
                 -b bin_198 \
                 -o SPLIT
```

Then I run the following to get the anvi'o estimate for genome completion:

```
anvi-estimate-genome-completeness -c SPLIT/bin_198/CONTIGS.db --concise

╒════════════╤══════════╤══════════════╤════════════════╤════════════════╤══════════════╤════════════════╕
│ bin name   │ domain   │   confidence │   % completion │   % redundancy │   num_splits │   total length │
╞════════════╪══════════╪══════════════╪════════════════╪════════════════╪══════════════╪════════════════╡
│ CONTIGS    │ BACTERIA │          0.7 │          74.65 │          25.35 │          265 │        5346483 │
╘════════════╧══════════╧══════════════╧════════════════╧════════════════╧══════════════╧════════════════╛
```

Just like Vincent described. For bin 198, this was the situation:

|Approach|Completion|Redundancy|
|:--|:--|:--|
|CheckM|100.00| 86.76|
|anvi'o|74.65|25.35|

This is the genome:

![image](https://user-images.githubusercontent.com/197307/177958187-24990755-1e51-433d-9bc8-3e2a9d7a7f74.png)

Very well assembled, looking beautifully complete in many ways, and had good number of SCGs resolving to the same taxon:

![image](https://user-images.githubusercontent.com/197307/177957859-8bc257ea-1fd5-499b-a860-7e86ce915901.png)

But what I realized later that there were some contigs that did not have any gene calls, which was the moment of eureka and it took some clicking around and inspecting contigs. Here is one contig that looks exactly the way it should in non-eukaryotic / non-viral / non-some-weird-stuff genomes in general:

![image](https://user-images.githubusercontent.com/197307/177958819-7258c3e2-c341-4ebc-9349-57c9d6f8012b.png)

And here is another one, that doesn't have a single gene despite its length!

![image](https://user-images.githubusercontent.com/197307/177958707-eefe2866-1dcc-4846-bbb1-e65225cc1100.png)

One of those situations where the scientist has to determine very quickly whether this is a Nobel-worthy discovery, or a bioinformatics artifact 🤔  Or, what we call in data-enabled microbiology, Friday :p

I checked my calendar, and sighed loudly that woke up Kiki. It *was* Friday today. So I decided to start from scratch and eliminate the latter explanation before giving Vincent the good news he deserved. I first exported the contigs as FASTA for the new bin,

```
anvi-export-contigs -c SPLIT/bin_198/CONTIGS.db -o bin_198.fa
```

Then generate a new contigs-db from it:

```
anvi-gen-contigs-database -f bin_198.fa -o bin_198.db
```

I quickly compared the number of genes found in the old and the new bin by exporting gene sequences

```
anvi-get-sequences-for-gene-calls -c bin_198.db -o new_genes_bin_198.fa

anvi-get-sequences-for-gene-calls -c SPLIT/bin_198/CONTIGS.db -o original_genes_bin_198.fa

grep '>' original_genes_bin_198.fa | wc -l
2175

grep '>' new_genes_bin_198.fa | wc -l
4614

o_O
-bash: o_O: command not found
```

I had my answer already: Something went wrong during the generation of the original contigs database, and it is missing many of the genes (in the case of `bin_198`, more than half of them) :(

But I continued to run HMMs:

```
anvi-run-hmms -c bin_198.db

anvi-run-scg-taxonomy -c bin_198.db
```

Then run the completion estimate:

```
anvi-estimate-genome-completeness -c bin_198.db --concise
╒════════════╤══════════╤══════════════╤════════════════╤════════════════╤══════════════╤════════════════╕
│ bin name   │ domain   │   confidence │   % completion │   % redundancy │   num_splits │   total length │
╞════════════╪══════════╪══════════════╪════════════════╪════════════════╪══════════════╪════════════════╡
│ bin_198    │ BACTERIA │            1 │            100 │           83.1 │          254 │        5346483 │
╘════════════╧══════════╧══════════════╧════════════════╧════════════════╧══════════════╧════════════════╛
```

So, there we have it. After re-running the gene caller by means of regenerating the anvi'o contigs-db, we get this:

|Approach|Completion|Redundancy|
|:--|:--|:--|
|CheckM|100.00| 86.76|
|anvi'o|100.00|83.10|

Tadaa, but a sad one. While this solves the mystery, it adds work on your plate, Vincent :/ I have no idea what went wrong with the gene calling step, but I can tell you that this is the first time I'm seeing this. Probably it is somewhere in the `anvi-gen-contigs-database` log files at the HPC. I will add this to my list of "weird things I've seen that explain poor completion estimates".

Sorry for the inconvenience, and thanks again for reporting it.

**Comment by meren** [comment]
And a totally irrelevant note:

![image](https://user-images.githubusercontent.com/197307/177965226-ff7828eb-86ed-4985-b437-67487849c49b.png)

:)

**Comment by vdenef** [comment]
Great! Thanks so much Meren! (Autocorrect suggested merengue:-)). Teach at
a field station but will get back to it asap! You all rock so amazingly!!!!

Vincent

**Comment by vdenef** [comment]
Reran building the contigs db etc and that did the trick. Happy that we could add to your weird things that can explain completion anomalies list.

We have installed the dev version. I will run that one going forward. When running the self test I realized it it hope (v7.1-dev) rather than 7.2?

Cheers,
Vincent
----------------------------------------------------------
Vincent Denef (he/him)
Associate Professor
Department of Ecology and Evolutionary Biology
University of Michigan
4060 Biological Sciences Building
1105 N. University
Ann Arbor, MI  48109-1085
Phone: +1 (734) 764 6481
http://www-personal.umich.edu/~vdenef/

**Comment by meren** [comment]
It is indeed `v7.1-dev`, and not `v7.2` :)
